src/Path/Core.hs

```diff
@@ -1,77 +1,89 @@
-{-# LANGUAGE DeriveTraversable, LambdaCase, RankNTypes, ScopedTypeVariables #-}
+{-# LANGUAGE DeriveAnyClass, DeriveGeneric, DeriveTraversable, FlexibleContexts, FlexibleInstances, LambdaCase, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TupleSections, TypeApplications, TypeOperators #-}
 module Path.Core where
 
-import Control.Monad (ap)
-import Path.Name
-import Path.Plicity
-import Path.Usage
-import Prelude hiding (pi)
-import Text.Trifecta.Rendering (Span)
-
-data Core a
-  = Var (Name a)
-  | Lam (Plicit (Maybe User)) (Scope a)
-  | Core a :$ Plicit (Core a)
+import           Control.Applicative (Alternative (..))
+import           Control.Effect
+import           Control.Effect.Carrier
+import           Control.Effect.Sum
+import           Control.Monad.Module
+import qualified Data.Set as Set
+import           GHC.Generics (Generic1)
+import           Path.Name
+import           Path.Scope
+import           Path.Syntax
+import           Path.Term
+import           Prelude hiding (pi)
+
+data Core f a
+  = Lam (Scope () f a)
+  | f a :$ f a
+  | Let (f a) (Scope () f a)
   | Type
-  | Pi (Plicit (Maybe User, Usage, Core a)) (Scope a)
-  | Hole Gensym
-  | Ann Span (Core a)
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
+  | Pi (f a) (Scope () f a)
+  deriving (Foldable, Functor, Generic1, HFunctor, Traversable)
+
+deriving instance (Eq   a, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Core f a)
+deriving instance (Ord  a, forall a . Eq   a => Eq   (f a)
+                         , forall a . Ord  a => Ord  (f a), Monad f) => Ord  (Core f a)
+deriving instance (Show a, forall a . Show a => Show (f a))          => Show (Core f a)
 
-newtype Scope a = Scope (Core (Incr a))
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
+instance Monad f => RModule (Core f) f where
+  Lam b   >>=* f = Lam (b >>=* f)
+  g :$ a  >>=* f = (g >>= f) :$ (a >>= f)
+  Let v b >>=* f = Let (v >>= f) (b >>=* f)
+  Type    >>=* _ = Type
+  Pi t b  >>=* f = Pi (t >>= f) (b >>=* f)
 
-instance Applicative Core where
-  pure = Var . Local
-  (<*>) = ap
+instance Syntax Core where
+  foldSyntax go bound free = \case
+    Lam b -> Lam (foldSyntax go bound free b)
+    f :$ a -> go free f :$ go free a
+    Let v b -> Let (go free v) (foldSyntax go bound free b)
+    Type -> Type
+    Pi t b -> Pi (go free t) (foldSyntax go bound free b)
 
-instance Monad Core where
-  a >>= f = joinT (fmap f a)
 
-lam :: Eq a => Plicit a -> Core a -> Core a
-lam (p :< n) b = Lam (p :< Nothing) (bind n b)
+lam :: (Eq a, Carrier sig m, Member Core sig) => a -> m a -> m a
+lam n b = send (Lam (bind1 n b))
```

Comment:
`fused-effects` 0.5 allows for `Scope` & therefore `Core` (and similar syntax functors) to be lifted into effectful contexts. Thus, we can use the same smart constructors to build `Term Core a`, `Term (Problem :+: Core) a`, and any other suitable `Core`-carrier.